src/Data/Error.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE ImplicitParams #-}
```

Comment:
This is the only place in the codebase where we use `ImplicitParams`. I'm assuming that it's necessary, but wanted to double-check. Also, the documentation for CallStack says this [here](https://hackage.haskell.org/package/base-4.17.2.0/docs/GHC-Stack-Types.html#t:CallStack):

```
NOTE: The intrepid user may notice that HasCallStack is just an alias for an implicit parameter ?callStack :: CallStack . This is an implementation detail and should not be considered part of the CallStack API, we may decide to change the implementation in the future.
```

I don't really know if this is something to be worried about, but I'd like to have someone who does give this a thumbs up

Reply:
So we can get rid of `ImplicitParams` by changing the code to be 

```
import GHC.Stack (CallStack, getCallStack, SrcLoc(..))

-- getSourceLocation returns SourceLocation with the filePath, line, col of the call site
getSourceLocation :: CallStack -> SourceLocation
getSourceLocation callStack = case getCallStack callStack of
  (_, loc) : _ -> SourceLocation (srcLocFile loc) (srcLocStartLine loc) (srcLocStartCol loc)
  _ -> SourceLocation "Unknown" 0 0
```

I know @csasarak had comments about this as well. So pulling him in to get his thoughts.

Reply:
When we spoke earlier, my comments were more that I think it's annoying to have to call `getSourceLocation` to get a source location, it'd be neat if it could just be automatic in order to make the work of future programmer's simpler. I don't know that any of my ideas are really much simpler than this though. If I'm reading it correct it's also optional to give `Errata` a location - you can just give it no blocks. Is that correct?

Regarding `ImplicitParams` itself, I think it's not ideal because it's relying on a global dynamic variable to get its info, but I also don't really know of a better way to do this. I would recommend using `HasCallStack` though instead of the `ImplicitParams` directly for the reasons Scott pointed out.

Reply:
Yea the location is optional, I was just using it in the areas where the errors cause our code to exit